src/OpenTelemetry.Instrumentation.Http/Implementation/HttpHandlerMetricsDiagnosticListener.cs

```diff
@@ -26,44 +26,110 @@ namespace OpenTelemetry.Instrumentation.Http.Implementation
     internal sealed class HttpHandlerMetricsDiagnosticListener : ListenerHandler
     {
         internal const string OnStopEvent = "System.Net.Http.HttpRequestOut.Stop";
+        private const string HttpClientMetricName = "http.client.duration";
 
         private readonly PropertyFetcher<HttpResponseMessage> stopResponseFetcher = new("Response");
         private readonly Histogram<double> httpClientDuration;
+        private readonly HttpClientInstrumentationMeterOptions options;
 
-        public HttpHandlerMetricsDiagnosticListener(string name, Meter meter)
+        public HttpHandlerMetricsDiagnosticListener(string name, Meter meter, HttpClientInstrumentationMeterOptions options)
             : base(name)
         {
-            this.httpClientDuration = meter.CreateHistogram<double>("http.client.duration", "ms", "measures the duration of the outbound HTTP request");
+            this.options = options;
+            this.httpClientDuration = meter.CreateHistogram<double>(HttpClientMetricName, "ms", "measures the duration of the outbound HTTP request");
         }
 
         public override void OnEventWritten(string name, object payload)
         {
             if (name == OnStopEvent)
             {
-                if (Sdk.SuppressInstrumentation)
-                {
-                    return;
-                }
+                this.OnStopActivity(Activity.Current, payload);
+            }
+        }
 
-                var activity = Activity.Current;
-                if (this.stopResponseFetcher.TryFetch(payload, out HttpResponseMessage response) && response != null)
-                {
-                    var request = response.RequestMessage;
+        private void OnStopActivity(Activity activity, object payload)
+        {
+            if (Sdk.SuppressInstrumentation)
+            {
+                return;
+            }
 
-                    TagList tags = default;
-                    tags.Add(new KeyValuePair<string, object>(SemanticConventions.AttributeHttpMethod, HttpTagHelper.GetNameForHttpMethod(request.Method)));
-                    tags.Add(new KeyValuePair<string, object>(SemanticConventions.AttributeHttpScheme, request.RequestUri.Scheme));
-                    tags.Add(new KeyValuePair<string, object>(SemanticConventions.AttributeHttpStatusCode, (int)response.StatusCode));
-                    tags.Add(new KeyValuePair<string, object>(SemanticConventions.AttributeHttpFlavor, HttpTagHelper.GetFlavorTagValueFromProtocolVersion(request.Version)));
-                    tags.Add(new KeyValuePair<string, object>(SemanticConventions.AttributeNetPeerName, request.RequestUri.Host));
+            if (!this.stopResponseFetcher.TryFetch(payload, out HttpResponseMessage response) || response == null)
+            {
+                // TODO: logging?
+                return;
+            }
+
+            var request = response.RequestMessage;
+            if (!this.TryFilterHttpRequestMessage(activity.OperationName, request))
```

Comment:
this looks like response message and not request?

Reply:
replace with response